File: Game Codes/exercise_blackjack.js
function initBlackJack() {
    const hitBtn = document.getElementById('hit-btn');
    const standBtn = document.getElementById('stand-btn');
    const doubleBtn = document.getElementById('double-btn');
    const betInput = document.getElementById('bet');
    const spinButton = document.getElementById('spinButton');
    const resetButton = document.getElementById('resetButton');
    const balanceDisplay = document.getElementById('balanceAmount');

    const suits = ["Hearts", "Diamonds", "Clubs", "Spades"];
    const ranks = ["2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"];
    const values = { "2": 2, "3": 3, "4": 4, "5": 5, "6": 6, "7": 7, "8": 8, "9": 9, "10": 10, "J": 10, "Q": 10, "K": 10, "A": 11 };

    let deck = [];
    let playerHand = [];
    let dealerHand = [];
    let currentBet = 0;
    let gameInProgress = false;

    function updateBalanceDisplay() {
        if (balanceDisplay && window.currency) {
            balanceDisplay.textContent = `${window.currency.getBalance().toFixed(2)}`;
        }
    }

    function displayOutput(message, clear = false) {
        const output = document.getElementById('output');
        if (clear) output.innerText = '';
        output.innerText += message + '\n';
        output.scrollTop = output.scrollHeight;
    }

    function displayDealer(message, clear = false) {
        const output = document.getElementById('dealer-hand');
        if (clear) output.innerText = '';
        output.innerText += message + '\n';
        output.scrollTop = output.scrollHeight;
    }

    function displayPlayer(message, clear = false) {
        const output = document.getElementById('player-hand');
        if (clear) output.innerText = '';
        output.innerText += message + '\n';
        output.scrollTop = output.scrollHeight;
    }

    function createDeck() {
        deck = [];
        for (let suit of suits) {
            for (let rank of ranks) {
                deck.push({ rank, suit });
            }
        }
        shuffleDeck();
    }

    function shuffleDeck() {
        for (let i = deck.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [deck[i], deck[j]] = [deck[j], deck[i]];
        }
    }

    function dealCard() {
        return deck.pop();
    }

    function calculateHandValue(hand) {
        let value = hand.reduce((sum, card) => sum + values[card.rank], 0);
        let aces = hand.filter(card => card.rank === "A").length;
        while (value > 21 && aces > 0) {
            value -= 10;
            aces--;
        }
        return value;
    }

    function displayHand(hand, isDealer = false, hideFirstCard = false) {
        if (isDealer) {
            if (hideFirstCard) {
                displayDealer(`Dealer's Hand: [Hidden], ${hand[1].rank} of ${hand[1].suit}`, true);
            } else {
                displayDealer(`Dealer's Hand: ${hand.map(card => `${card.rank} of ${card.suit}`).join(', ')} - Value: ${calculateHandValue(hand)}`, true);
            }
        } else {
            displayPlayer(`Your Hand: ${hand.map(card => `${card.rank} of ${card.suit}`).join(', ')} - Value: ${calculateHandValue(hand)}`, true);
        }
    }

    function placeBet() {
        if (gameInProgress) {
            displayOutput("Finish current game first!");
            return;
        }

        const bet = parseFloat(betInput.value, 10);
        if (isNaN(bet) || bet <= 0) {
            displayOutput("Please enter a valid bet.", true);
            return;
        }

        if (!currency.canAfford(bet)) {
            displayOutput("You don't have enough money to place that bet.", true);
            return;
        }

        if (!currency.placeBet(bet)) return;

        currentBet = bet;
        createDeck();
        playerHand = [dealCard(), dealCard()];
        dealerHand = [dealCard(), dealCard()];
        gameInProgress = true;

        displayHand(dealerHand, true, true);
        displayHand(playerHand);
        updateBalanceDisplay();

        if (calculateHandValue(playerHand) === 21) {
            determineWin();
        } else {
            enableButtons();
        }
    }

    function determineWin() {
        const playerValue = calculateHandValue(playerHand);
        const dealerValue = calculateHandValue(dealerHand);

        displayOutput("Final Results:");
        displayHand(playerHand);
        displayHand(dealerHand, true);

        if (playerValue > 21) {
            displayOutput("You busted! Dealer wins.", true);
        } else if (dealerValue > 21) {
            currency.addWinnings(currentBet * 2);
            displayOutput("Dealer busted! You win!", true);
        } else if (playerValue > dealerValue) {
            currency.addWinnings(currentBet * 2);
            displayOutput("You win!", true);
        } else if (playerValue < dealerValue) {
            displayOutput("Dealer wins!", true);
        } else {
            currency.addWinnings(currentBet);
            displayOutput("It's a tie!", true);
        }

        currentBet = 0;
        gameInProgress = false;
        updateBalanceDisplay();
        disableButtons();
    }

    function enableButtons() {
        hitBtn.disabled = false;
        standBtn.disabled = false;
        doubleBtn.disabled = false;
    }

    function disableButtons() {
        hitBtn.disabled = true;
        standBtn.disabled = true;
        doubleBtn.disabled = true;
    }

    function dealerTurn() {
        disableButtons();
        displayHand(dealerHand);
        while (calculateHandValue(dealerHand) < 17) {
            dealerHand.push(dealCard());
            displayHand(dealerHand);
        }
        determineWin();
    }

    function resetGame() {
        if (confirm("Don't give up! Press OK to get 10,000 and go ALL-IN!!!")) {
            currency.resetBalance();
            displayOutput("Game reset! Balance restored to $10,000", true);
            betInput.value = '';
            currentBet = 0;
            updateBalanceDisplay();
        }
    }

    function start() {
        displayOutput("Welcome to Blackjack\nPlace a bet to begin", true);
        updateBalanceDisplay();
        disableButtons();
    }

    // Event listeners
    spinButton.addEventListener('click', placeBet);
    resetButton.addEventListener('click', resetGame);
    hitBtn.addEventListener('click', function () {
        if (!gameInProgress) return;
        playerHand.push(dealCard());
        displayHand(playerHand);
        if (calculateHandValue(playerHand) > 21) {
            disableButtons();
            determineWin();
        }
    });
    standBtn.addEventListener('click', function () {
        if (!gameInProgress) return;
        dealerTurn();
    });
    doubleBtn.addEventListener('click', function () {
        if (!gameInProgress) return;
        if (currentBet * 2 <= currency.getBalance()) {
            if (currency.placeBet(currentBet)) {
                currentBet *= 2;
                playerHand.push(dealCard());
                displayHand(playerHand);
                if (calculateHandValue(playerHand) > 21) {
                    disableButtons();
                    determineWin();
                } else {
                    dealerTurn();
                }
                updateBalanceDisplay();
            }
        } else {
            displayOutput("Not enough funds to double!");
        }
    });

    start();
}

// ✅ Ensure everything waits for DOM + currency
document.addEventListener("DOMContentLoaded", function () {
    if (!window.currency) {
        document.addEventListener('currencyReady', initBlackJack);
    } else {
        initBlackJack();
    }
});

File: Game Codes/exercise_blackjack.py
import random

# Define deck and card values
suits = ["Hearts", "Diamonds", "Clubs", "Spades"]
ranks = ["2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"]
values = {"2": 2, "3": 3, "4": 4, "5": 5, "6": 6, "7": 7, "8": 8, "9": 9, "10": 10, 
          "J": 10, "Q": 10, "K": 10, "A": 11}  # Aces handled separately

# Function to create and shuffle deck
def create_deck():
    deck = [{"rank": rank, "suit": suit} for suit in suits for rank in ranks]
    random.shuffle(deck)
    return deck

# Function to deal a card
def deal_card(deck):
    return deck.pop()

# Function to calculate hand value, handling Aces
def calculate_hand_value(hand):
    value = sum(values[card["rank"]] for card in hand)
    aces = sum(1 for card in hand if card["rank"] == "A")

    while value > 21 and aces:
        value -= 10  # Convert an Ace from 11 to 1
        aces -= 1

    return value

# Function to display a hand
def display_hand(hand, hide_first_card=False):
    if hide_first_card:
        print("Dealer's Hand: [Hidden]", hand[1])
    else:
        print("Hand:", [f"{card['rank']} of {card['suit']}" for card in hand], "Value:", calculate_hand_value(hand))

# Function to handle bets
def place_bet(balance):
    while True:
        try:
            bet = int(input(f"You have ${balance}. Enter your bet: "))
            if 1 <= bet <= balance:
                return bet
            print("Invalid bet. Please enter an amount within your balance.")
        except ValueError:
            print("Please enter a valid number.")

# Function to play a hand
def play_hand(deck, player_hand, dealer_hand, balance, bet):
    # Player's turn
    while calculate_hand_value(player_hand) < 21:
        move = input("\nChoose: Hit (h), Stand (s), Double Down (d): ").lower()
        
        if move == 'h':
            player_hand.append(deal_card(deck))
            display_hand(player_hand)
        elif move == 'd':  # Double down
            if bet * 2 <= balance:
                bet *= 2
                player_hand.append(deal_card(deck))
                display_hand(player_hand)
                break  # Double down means you take one last card
            else:
                print("Not enough balance to double down!")
        elif move == 's':
            break
        else:
            print("Invalid choice, please enter 'h', 's', or 'd'.")

    # If player busts
    if calculate_hand_value(player_hand) > 21:
        print("\nYou busted! Dealer wins.")
        return -bet

    # Dealer's Turn
    print("\nDealer's turn:")
    display_hand(dealer_hand)

    while calculate_hand_value(dealer_hand) < 17:
        dealer_hand.append(deal_card(deck))
        display_hand(dealer_hand)

    # Determine winner
    player_value = calculate_hand_value(player_hand)
    dealer_value = calculate_hand_value(dealer_hand)

    print("\nFinal Results:")
    print("Your Hand:", player_hand, "Value:", player_value)
    print("Dealer's Hand:", dealer_hand, "Value:", dealer_value)

    if dealer_value > 21:
        print("Dealer busted! You win!")
        return bet
    elif player_value > dealer_value:
        print("You win!")
        return bet
    elif player_value < dealer_value:
        print("Dealer wins!")
        return -bet
    else:
        print("It's a tie!")
        return 0

# Function to handle splitting hands
def split_hand(deck, player_hand, dealer_hand, balance, bet):
    if player_hand[0]["rank"] == player_hand[1]["rank"]:
        choice = input("Do you want to split? (y/n): ").lower()
        if choice == 'y' and bet * 2 <= balance:
            hand1 = [player_hand[0], deal_card(deck)]
            hand2 = [player_hand[1], deal_card(deck)]
            print("\nPlaying first split hand:")
            display_hand(hand1)
            result1 = play_hand(deck, hand1, dealer_hand, balance, bet)

            print("\nPlaying second split hand:")
            display_hand(hand2)
            result2 = play_hand(deck, hand2, dealer_hand, balance, bet)

            return result1 + result2  # Return total earnings from both hands
    return play_hand(deck, player_hand, dealer_hand, balance, bet)

# Main game loop
def play_blackjack():
    balance = 1000  # Start with $1000
    
    while balance > 0:
        deck = create_deck()
        print("\nWelcome to Blackjack!")

        bet = place_bet(balance)

        # Initial hands
        player_hand = [deal_card(deck), deal_card(deck)]
        dealer_hand = [deal_card(deck), deal_card(deck)]

        print("\nYour hand:")
        display_hand(player_hand)
        display_hand(dealer_hand, hide_first_card=True)

        # Handle splitting
        earnings = split_hand(deck, player_hand, dealer_hand, balance, bet)

        balance += earnings
        print(f"\nYour new balance: ${balance}")

        if balance <= 0:
            print("You're out of money! Game over.")
            break

        play_again = input("\nPlay another round? (y/n): ").lower()
        if play_again != 'y':
            print("Thanks for playing! Final balance:", balance)
            break

# Run the game
play_blackjack()